pkg/scheduler/core/spreadconstraint/select_clusters.go

```diff
@@ -46,10 +52,99 @@ func selectBestClustersByCluster(spreadConstraint policyv1alpha1.SpreadConstrain
 		needCnt = totalClusterCnt
 	}
 
+	clusterInfos, ok := selectClustersByAvailableResource(groupClustersInfo.Clusters, int32(needCnt), needReplicas)
+	if !ok {
+		return nil, fmt.Errorf("no enough resource when selecting %d clusters", needCnt)
+	}
+
 	var clusters []*clusterv1alpha1.Cluster
-	for i := 0; i < needCnt; i++ {
-		clusters = append(clusters, groupClustersInfo.Clusters[i].Cluster)
+	for i := range clusterInfos {
+		clusters = append(clusters, clusterInfos[i].Cluster)
 	}
 
 	return clusters, nil
 }
+
+// if needClusterCount = 2, needReplicas = 80, member1 and member3 will be selected finally.
+// because the total resource of member1 and member2 is less than needReplicas although their scores is highest
+// --------------------------------------------------
+// | clusterName      | member1 | member2 | member3 |
+// |-------------------------------------------------
+// | score            |   60    |    50   |    40   |
+// |------------------------------------------------|
+// |AvailableReplicas |   40    |    30   |    60   |
+// |------------------------------------------------|
```

Comment:
This is an algorithm problem, can we abstract a common function to implement it and decoupling with the business logic?
@halfrost Would you like comment on it?

Reply:
First, let's discuss an abstract model for this problem. We want to find the maximum score that satisfies the following 2 conditions:
- 1. total availableReplicas >= needReplicas, and the smaller the total availableReplicas, the better, the less remaining. eg, needReplicas = 100, total availableReplicas = 101 is better than total availableReplicas = 110.
- 2. Only needClusterCount servers can be selected.

Your current solution does find a solution, but your solution is an approximate optimal solution, not a globally optimal solution. E.g, needClusterCount = 2, needReplicas = 80

```
-----------------------------------------------------------------------
| clusterName      | member1 | member2 | member3 |  member4 | member5 |
|----------------------------------------------------------------------
| score            |   60    |    50   |    40   |    40   |    40    |
|---------------------------------------------------------------------|
|AvailableReplicas |   40    |    30   |    35   |    100   |    40   |
|---------------------------------------------------------------------|
```

In this case, your code logic will find member 1 and member 4. score = 100, total availableReplicas = 140, but this is not a globally optimal solution, the optimal solution is member 1 and member 5 . score = 100, total availableReplicas = 80.

Then consider the time complexity of your solution. For your solution, the time complexity is O(m * n^2), where m is the number of times that selectClustersByAvailableResource function is called. n is the total number of candidateClusters.

I don't know whether you need to find an approximate optimal solution, or you must find a globally optimal solution.

If you must want to find the global optimal solution, this question should actually be answered with a two-dimensional 01 knapsack. The normal 01 backpack background you should already know. Here I will talk about the corresponding deformation of this question.

There are 2 conditions that need to be met here, the first condition is to find the minimum value larger than V. It can be known from the title condition that the value range larger than V is [v, total availableReplicas sum], and the second condition is to take needClusterCount items. We can use a dp array, dp[0...total sum][total count of cluster] to count all states.

Regarding time complexity, assuming that the number of clusters is a constant variable. At the beginning, calculate the dp data once when it is initialized, and the time complexity is O(n^2). After that, each query is to perform a for loop on the rows in this two-dimensional array. Start from v, find the first solution larger than v, and column = needClusterCount, which is the optimal solution. The final time complexity is O(n^2 + m*n), where m is the number of times that selectClustersByAvailableResource function is called. n is the total number of candidateClusters.

If the number of cluster does not change frequently, the above method is the best solution. If the number of cluster changes frequently, maybe need other optimization solution.

Reply:
here needs to find an approximate optimal solution.
1. the sum of selected availableReplicas >= needReplicas, but don't consider whether the total availableReplicas is small
   so E.g, needClusterCount = 2, needReplicas = 80
   the {member1, member4} is equal to {member1, member5}

Reply:
If you are only looking for approximate optimal solutions. Your solution can achieve this goal.

Reply:
My concern is the implementation is hard to understand.  Is there room to optimize?